# 🗣️ Behavioral & Project Storytelling — Turn Real Work Into Clear, Honest Stories

> **What you'll practise:** answering behavioral questions with STAR and CARL, building a reusable story bank, pitching a project in two minutes and surviving the drill-down questions, turning this course's capstones into interview stories, writing résumé bullets with measurable results, and asking good questions at the end. Two small, transparent tools help you: a **résumé-bullet checker** and a **story-bank coverage checker**.

| | |
|---|---|
| **Difficulty** | 🟢 Beginner → 🟡 Intermediate |
| **Time** | ~3 hours to read and run, +4–6 hours to write your own story bank and practise out loud |
| **Prerequisites** | None technically. You will get far more out of it after finishing at least one project, e.g. the [End-to-End ML Project](../13_Capstone_Projects/01_End_to_End_ML_Project.ipynb) capstone |
| **Tested with** | Python 3.12 · pandas 3.0 (only to display tables) · standard library `re` and `string` |
| **Interview relevance** | ⭐⭐⭐ Very high — almost every ML/AI engineer loop has a behavioral screen, a hiring-manager conversation, and a "walk me through a project" deep dive; a weak story can sink an otherwise strong technical loop |

## 🎯 How This Round Works

A **behavioral question** asks about something you actually did in the past: *"Tell me about a time you…"*. The idea behind it is simple: **how you handled real situations before is the best available predictor of how you will handle them again.** A **project deep dive** is the technical cousin: you describe one project and the interviewer keeps digging until they find the edge of what you know.

Think of it like a **cooking show judge**. The judge does not want the recipe you read online; they want to taste *your* dish and ask why you added each ingredient. Anyone can recite "I value teamwork" — the interviewer is listening for a specific moment, what *you* did, and what happened.

### Common formats

| Format | Who runs it | What it looks like | Typical length |
|---|---|---|---|
| **Recruiter / behavioral screen** | Recruiter or HR partner | "Tell me about yourself", "Why this role?", 1–2 light behavioral questions, logistics | 20–30 min |
| **Hiring-manager round** | Your future manager | Career story, motivation, 3–5 behavioral questions, how you work with people and handle ambiguity | 45–60 min |
| **Project deep dive** | A senior engineer or scientist | One project for the whole session: problem → data → metric → model → failures → deployment → "what would you change?" | 45–60 min |
| **Dedicated behavioral round** | A trained interviewer (e.g. Amazon's Leadership Principles questions) | Several "tell me about a time" questions with aggressive follow-ups | 45–60 min |
| **Behavioral slice inside a technical round** | Any interviewer | 5–10 minutes at the start: "tell me about a project you're proud of" | 5–10 min |

### What interviewers score

Most companies use a written rubric. The labels differ ("Leadership Principles", "values", "competencies") but they boil down to these:

| Signal | What "strong" sounds like | Red flags |
|---|---|---|
| **Ownership** | "I noticed X was broken, nobody owned it, so I…" | Everything was "we"; or "that wasn't my job" |
| **Impact** | A concrete result, ideally a number and who benefited | "It went well"; no result at all |
| **Collaboration** | Credits others precisely, explains how they worked together | Takes credit for the team's work — or can't say what *they* did |
| **Handling conflict** | Disagreed with data and respect, reached a decision, relationship intact | Blames or belittles a colleague; "I've never had a conflict" |
| **Handling ambiguity** | Clarified the goal, made assumptions explicit, shipped something small first | Waited for perfect requirements; or charged ahead without checking |
| **Learning from failure** | Owns the mistake, explains the fix *and* what changed afterwards | A disguised success ("I worked too hard"); blames circumstances |
| **Communication with non-technical people** | Translates metrics into business terms, uses examples, checks understanding | Jargon dumps; "they just didn't get it" |
| **Judgement & ethics** | Raises risks (bias, privacy, misuse) early and proposes options | Ships anything the metric rewards |

### Red flags that sink otherwise good candidates

- **Invented or borrowed stories.** Follow-ups ("what exactly did the dashboard show?") expose them within two questions — and it is dishonest. Every example answer in this notebook is **clearly labelled as an example**; use it to learn the *shape*, never as your content.
- **Inflated metrics.** "99% accuracy" on an imbalanced dataset, or "improved revenue by 40%" when you cannot explain how it was measured.
- **No "I".** The interviewer can't hire your team. Say "we" for context and "I" for your actions.
- **Rambling.** A 6-minute answer with the result buried at the end. Aim for **~2 minutes**, then let them ask follow-ups.
- **Negativity** about past employers, teammates, or interviewers.
- **No reflection.** Every story should end with what you learned or would do differently.

### How to use this notebook

| Section | What you'll do |
|---|---|
| [1. Answer Structures: STAR and CAR/CARL 🟢](#1.-Answer-Structures:-STAR-and-CAR/CARL-🟢) | Learn the formats; see a weak vs strong rewrite of the same story |
| [2. Building a Story Bank 🟢](#2.-Building-a-Story-Bank-🟢) | Map 8–10 real stories to competencies; check coverage with code |
| [3. The Project Deep Dive 🟡](#3.-The-Project-Deep-Dive-🟡) | A 2-minute pitch structure and how to handle drill-down questions |
| [4. Turning This Course's Capstones Into Stories 🟡](#4.-Turning-This-Course's-Capstones-Into-Stories-🟡) | Pitch skeletons and likely follow-ups for each capstone |
| [5. Common Behavioral Questions 🟡](#5.-Common-Behavioral-Questions-🟡) | 30 questions with hidden example answers |
| [6. Writing Résumé Bullets for ML Projects 🟢](#6.-Writing-Résumé-Bullets-for-ML-Projects-🟢) | Action + method + measurable result; a bullet checker |
| [7. Questions to Ask and Follow-Up Etiquette 🟢](#7.-Questions-to-Ask-and-Follow-Up-Etiquette-🟢) | End the interview well |
| [🏋️ Timed Drills](#🏋️-Timed-Drills) · [📋 Answer Framework & Cheat Sheet](#📋-Answer-Framework-&-Cheat-Sheet) · [📚 Resources](#📚-Resources) · [➡️ What's Next](#➡️-What's-Next) | 3 mock sets with a scoring rubric, one-page templates, checklist |

**Practise out loud.** Reading a good answer feels like knowing it; saying it with a timer running is a different skill. Open each hidden answer only *after* you have answered yourself.

## ⚙️ Setup

This notebook needs only the Python standard library plus pandas for displaying tables. If pandas is missing, uncomment the `%pip` line, run it once, and restart the kernel.

The `check()` helper gives instant feedback on the exercises: **✅** correct, **⏳** not attempted yet, **❌** wrong (with a hint).

In [1]:
# %pip install -q "pandas>=2.2"

import re
import string
import sys
from collections import Counter

import pandas as pd

print(f"Python {sys.version.split()[0]} | pandas {pd.__version__}")
pd.set_option("display.max_colwidth", 80)
pd.set_option("display.width", 160)


def check(name, got, expected, hint=""):
    """✅ if correct, ⏳ if not attempted yet (None / ...), ❌ AssertionError with a hint otherwise."""
    if got is None or got is ...:
        print(f"⏳ {name}: not attempted yet — replace None with your answer.")
        return
    assert got == expected, f"❌ {name}: got {got!r}, expected {expected!r}. {hint}"
    print(f"✅ {name}: correct!")

Python 3.12.11 | pandas 3.0.5


## 1. Answer Structures: STAR and CAR/CARL 🟢

Without a structure, stories wander: background for three minutes, then "…and it worked out". A structure is a **coat hanger** — the same shape every time, so the listener always knows where they are.

| Format | Parts | Best for |
|---|---|---|
| **STAR** | **S**ituation → **T**ask → **A**ction → **R**esult | The default. Most interviewers and many company prep guides use these words |
| **CAR** | **C**ontext (situation + task in one) → **A**ction → **R**esult | Shorter answers, phone screens, résumé bullets |
| **CARL** | **C**ontext → **A**ction → **R**esult → **L**earning | Failure and conflict questions, where reflection is half the score |

**Time budget for a ~2-minute answer:**

```
Situation + Task  ███░░░░░░░░░░░░░░░░░   ~15–20%   "just enough context to follow"
Action            ████████████░░░░░░░░   ~50–60%   "what YOU did, step by step, and why"
Result            ███░░░░░░░░░░░░░░░░░   ~15%      "what changed — a number if you have one"
Learning          ██░░░░░░░░░░░░░░░░░░   ~10%      "what you'd repeat or do differently"
```

The most common mistake is flipping that budget: long situation, one-line action, no result.

### Weak vs strong: the same story, rewritten

*Both versions below are an **illustrative example** written for this notebook — not a real project.*

**❌ Weak** (question: *"Tell me about a time you improved something"*)

> "So at my internship we had this forecasting pipeline and it was really slow, like really slow, and everyone complained about it. We looked into it and there were a lot of problems with the code. We ended up refactoring a lot of things and using better pandas functions, and it was a team effort. In the end it was much faster and people were happy. I learned a lot about pandas."

**✅ Strong (STAR + a learning)**

> **Situation:** "During my internship, the team's weekly sales-forecast pipeline took about six hours, so analysts received numbers a day late."
> **Task:** "My lead asked me to get it under an hour without changing the forecasts."
> **Action:** "I profiled it first instead of guessing, and found most of the time was in a row-by-row `apply`. I rewrote that step as a vectorized group-by. Because the output fed finance reports, I also wrote a regression test that compared old and new outputs on three months of history before switching over, and I walked the analyst who owned the report through the diff."
> **Result:** "The run dropped to about 40 minutes, outputs matched exactly, and analysts got forecasts the same morning."
> **Learning:** "Profiling before optimizing saved me from rewriting the parts that weren't slow — I now profile first by default."

Why the second one wins: a clear goal and constraint, "I" for actions, a *reason* behind each action, a verification step, a concrete result, and a reflection.

A quick, transparent signal you can compute on your own written drafts: **how often you say "I" vs "we"**, and whether the story contains any number at all. It is a *nudge*, not a grade — a great teamwork story legitimately has plenty of "we".

In [2]:
WEAK_STORY = (
    "So at my internship we had this forecasting pipeline and it was really slow, like really slow, and everyone "
    "complained about it. We looked into it and there were a lot of problems with the code. We ended up refactoring "
    "a lot of things and using better pandas functions, and it was a team effort. In the end it was much faster and "
    "people were happy. I learned a lot about pandas."
)
STRONG_STORY = (
    "During my internship, the team's weekly sales-forecast pipeline took about six hours, so analysts received "
    "numbers a day late. My lead asked me to get it under an hour without changing the forecasts. I profiled it first "
    "instead of guessing, and found most of the time was in a row-by-row apply. I rewrote that step as a vectorized "
    "group-by. Because the output fed finance reports, I also wrote a regression test that compared old and new "
    "outputs on three months of history before switching over, and I walked the analyst who owned the report through "
    "the diff. The run dropped to about 40 minutes, outputs matched exactly, and analysts got forecasts the same "
    "morning. Profiling before optimizing saved me from rewriting the parts that weren't slow."
)

NUMBER_WORDS = {"one", "two", "three", "four", "five", "six", "seven", "eight", "nine", "ten", "twice", "half", "hundred", "thousand"}


def story_signals(text):
    """Count ownership pronouns, filler words, and concrete quantities in a written story draft."""
    words = re.findall(r"[a-z']+|\d+", text.lower())
    counts = Counter(words)
    return {
        "words": len(words),
        "I / my": counts["i"] + counts["my"] + counts["me"],
        "we / our": counts["we"] + counts["our"] + counts["us"],
        "fillers (so, like, really, a lot)": counts["so"] + counts["like"] + counts["really"] + text.lower().count("a lot"),
        "quantities": sum(1 for w in words if w.isdigit() or w in NUMBER_WORDS),
    }


pd.DataFrame({"weak": story_signals(WEAK_STORY), "strong": story_signals(STRONG_STORY)})

,weak,strong
words,72,130
I / my,2,8
we / our,3,0
"fillers (so, like, really, a lot)",7,1
quantities,0,3


The table is computed from the two texts above: the strong version is longer but carries its weight in *actions* ("I profiled…", "I rewrote…", "I walked…") and concrete quantities, while the weak version leans on "we" and fillers.

### ✍️ Your Turn

Label each sentence of this *example* story as `"S"`, `"T"`, `"A"`, or `"R"` (Situation, Task, Action, Result). Put the labels in a list, in order.

In [3]:
example_story = [
    "Our churn model's weekly scores stopped matching what the retention team saw in their CRM.",       # 0
    "As the person who built the pipeline, I had to find the cause before the next campaign on Friday.",  # 1
    "I compared feature distributions between training data and the last four weekly batches.",           # 2
    "That showed a renamed column was silently filled with zeros, so I added a schema check that fails the job.",  # 3
    "Scores matched again for the Friday campaign, and the check has since caught two more upstream changes.",    # 4
]
star_labels = None  # TODO: e.g. ["S", "T", ...]
check("star_labels", star_labels, ["S", "T", "A", "A", "R"],
      hint="Sentence 0 sets the scene, 1 is your responsibility, the middle sentences are what you did, the last is the outcome.")

⏳ star_labels: not attempted yet — replace None with your answer.


<details><summary>💡 Show solution</summary>

```python
star_labels = ["S", "T", "A", "A", "R"]
check("star_labels", star_labels, ["S", "T", "A", "A", "R"])
```

Sentence 3 contains both a finding and a fix — both are things *you did*, so it's Action. Notice the Result has two parts: the immediate outcome (Friday campaign) and the lasting one (the check keeps catching problems). Lasting impact is gold in behavioral answers.
</details>

> 💡 **Interview angle:** Interviewers rarely say "use STAR", but many score against it silently: *Was the situation clear? What did the candidate personally do? What was the measurable result?* If you forget the result, a good interviewer asks "and what happened?" — a weaker one just writes "no impact shown".

## 2. Building a Story Bank 🟢

You cannot invent a good story in 10 seconds under pressure. So you prepare a **story bank**: 8–10 real experiences written down *once*, each tagged with the competencies it can demonstrate. In the interview you just pick the best-fitting story — like a musician choosing from a practised set list instead of composing on stage.

**Where stories come from** (early-career and career-switchers have more than they think):
- internships, jobs (including non-tech jobs — conflict and deadlines are universal), research, teaching assistant work
- course capstones and personal projects — especially the parts that *went wrong*
- open-source contributions, hackathons, Kaggle competitions, study groups, volunteering

### The template

Write each story as one row. Keep the numbers you remember **exactly as they were** — if you don't know a number, write what you *do* know ("the job went from failing weekly to not failing in the two months I was there").

| Field | What to write | Example (illustrative) |
|---|---|---|
| **Title** | A 3–6 word handle you'll remember | "Silent zero-filled column" |
| **Context** | 1–2 sentences: where, when, what was at stake | "Internship; churn scores fed a weekly retention campaign" |
| **Your actions** | 3–5 bullets of what *you* did and why | "Compared feature distributions; found renamed column; added schema check" |
| **Result** | What changed, with a number if one exists | "Campaign ran on time; check caught 2 later upstream changes" |
| **Learning** | What you'd repeat or change | "Validate inputs at pipeline boundaries, not just model outputs" |
| **Competencies** | Which question types it can answer | ownership, incomplete data, learning from failure |

### Reusing one story for several questions

One strong story can answer 3–4 different questions **by changing the emphasis**, not the facts:

| Question | Same story, different spotlight |
|---|---|
| "Tell me about a time you took ownership" | Nobody owned data validation; you noticed and fixed it |
| "Tell me about a mistake" | You shipped the pipeline without input checks in the first place |
| "A decision with incomplete data" | You had to decide by Friday whether to trust the scores |
| "Explaining something technical" | How you explained "silent zero-filling" to the retention manager |

Aim for **every competency covered by at least two stories**, so you never have to use the same story twice with the same interviewer panel.

### The story-bank coverage checker

The code below turns a story bank into a **coverage matrix** and reports gaps. The competency list matches the question groups in [section 5](#5.-Common-Behavioral-Questions-🟡). The example bank uses **placeholder stories** — replace them with yours.

In [4]:
COMPETENCIES = [
    "ownership", "impact", "collaboration", "conflict", "ambiguity", "failure",
    "non_technical_communication", "learning_quickly", "ethics_fairness",
    "incomplete_data_decision", "pushing_back", "deadline_pressure",
]

# Placeholder bank — titles describe the *kind* of story, not real events. Replace every row with your own.
example_bank = [
    {"title": "Speeding up a slow data pipeline", "competencies": ["ownership", "impact"]},
    {"title": "Silent zero-filled column in production", "competencies": ["ownership", "incomplete_data_decision", "deadline_pressure"]},
    {"title": "Disagreement over the evaluation metric", "competencies": ["conflict", "pushing_back", "collaboration"]},
    {"title": "Vague 'add AI search' request", "competencies": ["ambiguity", "non_technical_communication"]},
    {"title": "Learning PyTorch for a hackathon", "competencies": ["learning_quickly", "deadline_pressure"]},
    {"title": "Explaining precision vs recall to ops", "competencies": ["non_technical_communication", "impact"]},
    {"title": "Group capstone with an absent teammate", "competencies": ["collaboration", "conflict"]},
    {"title": "Choosing a model with only 3 weeks of data", "competencies": ["incomplete_data_decision", "ambiguity", "learning_quickly"]},
]


def coverage_report(bank, competencies=COMPETENCIES, min_stories=2):
    """Return a coverage matrix plus gaps (0 stories), thin spots (< min_stories) and unknown tags (likely typos)."""
    known = set(competencies)
    unknown = sorted({tag for story in bank for tag in story["competencies"] if tag not in known})
    matrix = pd.DataFrame(
        [["✓" if comp in story["competencies"] else "" for comp in competencies] for story in bank],
        index=[story["title"] for story in bank], columns=competencies,
    )
    counts = {comp: sum(comp in story["competencies"] for story in bank) for comp in competencies}
    gaps = [comp for comp, n in counts.items() if n == 0]
    thin = [comp for comp, n in counts.items() if 0 < n < min_stories]
    return {"matrix": matrix, "counts": counts, "gaps": gaps, "thin": thin, "unknown_tags": unknown}


report = coverage_report(example_bank)
print("❌ gaps (no story at all):", report["gaps"])
print("⚠️ thin (only one story) :", report["thin"])
print("❓ unknown tags (typos?)  :", report["unknown_tags"])
report["matrix"]

❌ gaps (no story at all): ['failure', 'ethics_fairness']
⚠️ thin (only one story) : ['pushing_back']
❓ unknown tags (typos?)  : []


,ownership,impact,collaboration,conflict,ambiguity,failure,non_technical_communication,learning_quickly,ethics_fairness,incomplete_data_decision,pushing_back,deadline_pressure
Speeding up a slow data pipeline,✓,✓,,,,,,,,,,
Silent zero-filled column in production,✓,,,,,,,,,✓,,✓
Disagreement over the evaluation metric,,,✓,✓,,,,,,,✓,
Vague 'add AI search' request,,,,,✓,,✓,,,,,
Learning PyTorch for a hackathon,,,,,,,,✓,,,,✓
Explaining precision vs recall to ops,,✓,,,,,✓,,,,,
Group capstone with an absent teammate,,,✓,✓,,,,,,,,
Choosing a model with only 3 weeks of data,,,,,✓,,,✓,,✓,,


Read the output: the example bank has **no story** for `failure` or `ethics_fairness` — two of the most frequently asked question types — and only one for `pushing_back`. That is precisely the kind of gap you want to find *before* the interview.

**Why a coverage check instead of just "more stories"?** Candidates naturally collect success stories. The checker forces you to find a real failure story and a real judgement-call story, which are the ones people freeze on.

### ✍️ Your Turn

Add story entries to `new_stories` so that `coverage_report(example_bank + new_stories)` has **no gaps**. In your real bank, each entry must be a true experience — here, a descriptive placeholder title is enough to make the checker pass.

In [5]:
new_stories = None  # TODO: a list of dicts like {"title": "...", "competencies": ["..."]}
filled = None if new_stories is None else coverage_report(example_bank + new_stories)["gaps"]
check("coverage gaps filled", filled, [],
      hint="Look at report['gaps'] above and tag one or two new stories with exactly those competency names.")

⏳ coverage gaps filled: not attempted yet — replace None with your answer.


<details><summary>💡 Show solution</summary>

```python
new_stories = [
    # Placeholders — each must become one of YOUR real experiences
    {"title": "Model I shipped that underperformed", "competencies": ["failure", "ownership"]},
    {"title": "Spotting a biased label source", "competencies": ["ethics_fairness", "pushing_back"]},
]
filled = coverage_report(example_bank + new_stories)["gaps"]
check("coverage gaps filled", filled, [])
print("thin now:", coverage_report(example_bank + new_stories)["thin"])
```

Pairing `ethics_fairness` with `pushing_back` is natural: raising a fairness concern usually means disagreeing with a plan. The `thin` list shows what to strengthen next.
</details>

> 💡 **Interview angle:** "Tell me about a time…" questions often arrive back-to-back from different interviewers who compare notes afterwards. A story bank with ≥2 stories per competency means you never repeat a story to the same panel — and never have to improvise one.

## 3. The Project Deep Dive 🟡

In a deep dive the interviewer picks **one project** (usually the top one on your résumé) and spends 30–60 minutes on it. They want to know three things: *Did you really do this? Do you understand why each decision was made? Can you reason about what you'd do differently?*

### The 2-minute pitch

Open with a crisp overview so the interviewer can choose where to dig. Six beats, in order:

| # | Beat | Question it answers | Share of 2 min |
|---|---|---|---|
| 1 | **Problem** | What was the goal and who cared? | ~15 s |
| 2 | **Constraints** | Data size, labels, latency, budget, deadline, privacy | ~15 s |
| 3 | **Approach** | Baseline first, then the model/system you built and why | ~30 s |
| 4 | **Trade-offs** | What you chose *not* to do, and the cost of your choice | ~20 s |
| 5 | **Results with metrics** | Offline metric vs a baseline, and any real-world effect | ~20 s |
| 6 | **What you'd change** | Honest limitations and next steps | ~20 s |

People speak at very roughly **130–160 words per minute** in conversation, so a 2-minute pitch is about **250–300 words**. The helper below estimates speaking time from a written draft (at 140 wpm by default — time yourself to find your own pace).

In [6]:
def speaking_seconds(text, words_per_minute=140):
    """Estimate how long a written answer takes to say out loud."""
    n_words = len(re.findall(r"\S+", text))
    return n_words, round(n_words / words_per_minute * 60)


EXAMPLE_PITCH = """\
Problem: a subscription business wanted to know which customers were likely to cancel next month, so the retention team could call the riskiest ones first.
Constraints: about seven thousand customers with a churn rate near one in four, tabular data only, and the team could only call a few hundred people a week, so ranking quality at the top mattered more than overall accuracy.
Approach: I started with a majority-class baseline and a logistic regression, then tried gradient boosting inside a scikit-learn pipeline so preprocessing was fitted on training folds only. I tuned on a validation split and touched the test set once.
Trade-offs: boosting was a little better than logistic regression but harder to explain, so I kept both and reported feature effects with permutation importance. I chose ROC-AUC plus precision in the top decile instead of accuracy, because accuracy looks great on imbalanced data even for a useless model.
Results: on the held-out test set the boosted model beat the logistic regression baseline on both metrics, and I served it behind a small FastAPI endpoint with input validation and a test suite.
What I'd change: the data is a single snapshot, so I can't measure drift; next I'd do a time-based split and estimate the value of a call against its cost to pick the threshold.
"""
n_words, seconds = speaking_seconds(EXAMPLE_PITCH)
verdict = "fits in 2 minutes ✅" if seconds <= 120 else "too long — cut it ✂️"
print(f"{n_words} words ≈ {seconds} s at 140 wpm → {verdict}")

218 words ≈ 93 s at 140 wpm → fits in 2 minutes ✅


*The pitch above is an illustrative example of the **shape**. It deliberately says "beat the baseline" instead of quoting numbers — your pitch must quote **your own** measured numbers (see section 4).*

### ✍️ Your Turn

The beats of another *example* pitch got shuffled. Put the keys in the correct pitch order (problem → constraints → approach → trade-offs → results → what you'd change).

In [7]:
shuffled_beats = {
    "a": "I'd add a reranker next, because most wrong answers came from the right chunk ranked fifth.",
    "b": "Support agents spent minutes searching a 400-page policy manual for each ticket.",
    "c": "Hybrid BM25 + embedding retrieval beat embeddings alone on recall@5 on my 60-question test set.",
    "d": "I built a baseline keyword search first, then a RAG pipeline with chunking, hybrid retrieval and citations.",
    "e": "The manual could not leave the company network, so everything had to run on a local model.",
    "f": "I chose smaller chunks for more precise citations, accepting that some answers needed two chunks.",
}

In [8]:
pitch_order = None  # TODO: a list of the six keys, e.g. ["b", ...]
check("pitch_order", pitch_order, ["b", "e", "d", "f", "c", "a"],
      hint="Problem = the pain; constraints = what limited you; results mention a metric; the last beat looks forward.")

⏳ pitch_order: not attempted yet — replace None with your answer.


<details><summary>💡 Show solution</summary>

```python
pitch_order = ["b", "e", "d", "f", "c", "a"]
check("pitch_order", pitch_order, ["b", "e", "d", "f", "c", "a"])
print("\n".join(shuffled_beats[k] for k in pitch_order))
```

Notice beat "f" names the **cost** of the choice ("some answers needed two chunks"). A trade-off without a cost is just a decision; naming the cost is what signals seniority.
</details>

### Handling drill-down questions

After the pitch, the interviewer digs. Almost every follow-up falls into one of these buckets — prepare one honest sentence for each **before** the interview:

| Bucket | Typical question | What they're checking | Prepare |
|---|---|---|---|
| **Data** | "Where did the labels come from? How clean were they?" | You looked at the data, not just `.fit()` | Size, source, label quality, one surprising thing you found |
| **Leakage & splits** | "How did you split? Could information leak?" | You understand evaluation validity | Split strategy and *why* (random / time / group) |
| **Metrics** | "Why that metric?" | You connected the metric to the goal | Metric ↔ business cost; why not accuracy |
| **Baselines** | "Compared to what?" | Your number means something | The simplest baseline and its score |
| **Failures** | "What didn't work?" | Honesty and learning | Two things you tried that failed, and why |
| **Scale** | "What if data were 100× bigger / latency 10× tighter?" | You can reason past the prototype | The first bottleneck and one concrete fix |
| **Ownership** | "Which parts did *you* do?" | Team project vs your contribution | A precise split of who did what |

**The golden rule:** if you don't know, say so and reason out loud — *"I didn't measure that. My guess is X because Y, and I'd check it by Z."* Interviewers respect that far more than a confident guess they can disprove.

**Q1. "Walk me through your project." How do you start, and how long should you talk before pausing?**

<details><summary>Show answer</summary>

- **30-second answer:** Give the six-beat overview in about two minutes — problem, constraints, approach, trade-offs, results with a baseline, what you'd change — then stop and ask *"Which part would you like me to go deeper on?"* That hands control back and shows you can summarize.
- **Go deeper:** Tailor the emphasis to the interviewer: a hiring manager wants impact and collaboration, an ML engineer wants data, evaluation and deployment. Bring one simple diagram in your head (data → features → model → serving → monitoring) you could sketch if asked.
- **❌ Common wrong answer:** Starting with the tech stack ("I used PyTorch, Docker, and Kubernetes…") and talking for ten minutes before stating the problem.

</details>

**Q2. "Why did you choose that metric?"**

<details><summary>Show answer</summary>

- **30-second answer:** Tie the metric to the decision the model supports. *Example:* "The team could only act on the top few hundred predictions each week, so I optimized ranking at the top — precision in the top decile — and tracked ROC-AUC as a threshold-free summary. Accuracy was misleading because the majority class alone scores high."
- **Go deeper:** Mention the cost asymmetry (false negatives vs false positives), how you picked a threshold on the validation set, and what online metric would confirm the offline one.
- **❌ Common wrong answer:** "Accuracy, because it's the standard metric" — on imbalanced data a model that never predicts the minority class can look excellent.

</details>

**Q3. "What was your baseline, and how much better was your model?"**

<details><summary>Show answer</summary>

- **30-second answer:** Name the simplest reasonable baseline (majority class, mean prediction, a rule, logistic regression, BM25 for search) and state both numbers on the **same** held-out data. If the gain was small, say so and explain why the extra complexity was or wasn't worth it.
- **Go deeper:** Mention variance: with a small test set, a 1-point difference may be noise — cross-validation spread or a bootstrap interval shows whether the gain is real.
- **❌ Common wrong answer:** "I didn't have a baseline, but 87% is good." A number without a comparison can't be judged.

</details>

**Q4. "What didn't work?"**

<details><summary>Show answer</summary>

- **30-second answer:** Give one or two genuine dead ends, what you learned from each, and how that changed your approach. *Example:* "A deeper network overfit badly on 2,000 rows; the validation loss told me within a few epochs, so I went back to gradient boosting and spent the time on features instead."
- **Go deeper:** The best failure answers show a **fast feedback loop** — you noticed quickly because you had a validation set, a baseline, and logging.
- **❌ Common wrong answer:** "Everything worked as planned." It sounds either untrue or like the project was too easy.

</details>

**Q5. "How would this work at 100× the data or traffic?"**

<details><summary>Show answer</summary>

- **30-second answer:** Identify the first bottleneck rather than listing technologies: *"Training on one machine still fits at 100× for this tabular model, but feature computation in pandas wouldn't — I'd move it to SQL/Spark. For serving, batch scoring nightly is cheaper than an API if predictions don't need to be real-time."*
- **Go deeper:** Discuss what changes besides compute: drift monitoring, retraining schedule, data validation, latency budgets, cost per prediction, and for LLM apps caching and token cost.
- **❌ Common wrong answer:** "Put it on Kubernetes" with no bottleneck analysis.

</details>

> 💡 **Interview angle:** Deep dives are where résumé inflation gets caught. The interviewer keeps asking "why?" until you hit the edge of your understanding — so only put projects on your résumé that you can defend for 45 minutes, and know your baseline, metric, split and biggest failure cold.

## 4. Turning This Course's Capstones Into Stories 🟡

Module 13 has four capstones. Each is a ready-made deep-dive project **once you have built it yourself** and measured your own results. The notebooks are still being finished, so the skeletons below describe each project at the level that stays true regardless of the exact dataset or library version:

| Capstone | What it is (in one line) |
|---|---|
| [01 · End-to-End ML Project](../13_Capstone_Projects/01_End_to_End_ML_Project.ipynb) | A tabular prediction problem taken from raw data to a tested, served model |
| [02 · End-to-End DL Project](../13_Capstone_Projects/02_End_to_End_DL_Project.ipynb) | A deep-learning model trained in PyTorch with proper validation, exported and served |
| [03 · LLM RAG Assistant Project](../13_Capstone_Projects/03_LLM_RAG_Assistant_Project.ipynb) | A document question-answering assistant with retrieval, citations and an evaluation set |
| [04 · AI Agent Project](../13_Capstone_Projects/04_AI_Agent_Project.ipynb) | A tool-calling agent with guardrails, tracing and task-level evaluation |

### ⚠️ The one rule: your numbers, or no numbers

Everything in `{curly braces}` below is a **blank you fill from your own run**. Never copy a number from a tutorial, from someone else's repo, or from this notebook. If you didn't measure something, say "I didn't measure that" — then say how you would.

To make that rule mechanical, `render_pitch()` fills a template from a dictionary of facts and **refuses to produce a pitch while any blank is missing** — it will never invent a value for you.

In [9]:
def template_fields(template):
    """Names of all {placeholders} in a template, in order of first appearance."""
    seen = []
    for _, field, _, _ in string.Formatter().parse(template):
        if field and field not in seen:
            seen.append(field)
    return seen


def render_pitch(template, facts):
    """Fill the template only if every blank has a real value; otherwise list what is missing."""
    missing = [f for f in template_fields(template) if facts.get(f) in (None, "")]
    if missing:
        print(f"⏳ Not ready — measure/fill these first: {missing}")
        return None
    return template.format(**facts)


ML_PITCH = (
    "Problem: {who} needed to predict {target} so they could {decision}.\n"
    "Constraints: {n_rows} rows, {positive_rate} positive class, {constraint}.\n"
    "Approach: baseline {baseline_model}, then {final_model} in a leak-free pipeline, tuned on a validation split.\n"
    "Trade-offs: I chose {tradeoff_choice} over {tradeoff_alternative} because {tradeoff_reason}.\n"
    "Results: on the untouched test set, {metric} was {final_score} vs {baseline_score} for the baseline; served via {serving}.\n"
    "What I'd change: {next_step}."
)

print("Blanks in the ML pitch:", template_fields(ML_PITCH))
my_facts = {"who": "a retention team", "target": "which customers will cancel", "decision": "call the riskiest first",
            "baseline_model": "logistic regression", "final_model": "gradient boosting", "serving": "a FastAPI endpoint"}
pitch = render_pitch(ML_PITCH, my_facts)       # the qualitative parts are filled, the measured ones are not

Blanks in the ML pitch: ['who', 'target', 'decision', 'n_rows', 'positive_rate', 'constraint', 'baseline_model', 'final_model', 'tradeoff_choice', 'tradeoff_alternative', 'tradeoff_reason', 'metric', 'final_score', 'baseline_score', 'serving', 'next_step']
⏳ Not ready — measure/fill these first: ['n_rows', 'positive_rate', 'constraint', 'tradeoff_choice', 'tradeoff_alternative', 'tradeoff_reason', 'metric', 'final_score', 'baseline_score', 'next_step']


That refusal is the point: the qualitative facts are known before you start, but `n_rows`, `positive_rate`, `final_score` and `baseline_score` only exist after **you** run the project. Paste this template into your own notes, fill it from your results, and time it with `speaking_seconds()` from section 3.

Below, each capstone has a pitch skeleton and the follow-ups interviewers are most likely to ask. Hidden under each is what a good answer covers.

### 🧮 Capstone 01 — End-to-End ML Project

**Pitch skeleton:** *"{Who} needed to predict {target} to {decision}. The data had {n_rows} rows with {class balance / target distribution}. I built a {baseline} baseline, then a {final model} inside a scikit-learn pipeline so preprocessing was fitted only on training data. I picked {metric} because {why it matches the decision}. On the held-out test set it reached {your score} vs {baseline score}. I explained predictions with {method}, packaged the model behind {serving}, and wrote tests for {what}. Next I'd {one honest improvement}."*

**Likely follow-ups:**
1. How did you split the data, and how do you know nothing leaked?
2. Why this metric? What threshold would you use, and who decides it?
3. Which features mattered most — and could any of them be a proxy for the label or a sensitive attribute?
4. How much better than the baseline was it, and is that difference bigger than the noise?
5. What happens when the input distribution shifts next quarter?
6. How does the served model get the *same* preprocessing as training?

<details><summary>What good answers cover</summary>

- **Split & leakage:** stratified (or time-based if rows have dates) split made *before* any fitting; imputers/scalers/encoders inside the `Pipeline`; features that exist only after the outcome removed.
- **Metric & threshold:** metric tied to the cost of errors; threshold picked on validation data from a precision/recall trade-off, owned by the business team.
- **Feature importance:** permutation importance on held-out data rather than only built-in tree importances; a sanity check for suspiciously strong features.
- **Noise:** cross-validation spread or a bootstrap interval on the test metric.
- **Drift:** monitor input distributions and prediction rates; retrain on a schedule or trigger; see the drift notebook in module 09.
- **Train/serve skew:** save the whole fitted pipeline (not just the model) and test the API against training-time predictions.

</details>

### 🖼️ Capstone 02 — End-to-End DL Project

**Pitch skeleton:** *"The goal was to classify {inputs} into {classes} for {use}. I had {n examples} labelled examples, split into train/validation/test. I started from {baseline — e.g. a small model or a pretrained backbone} and trained with {augmentation, optimizer, schedule}, using the validation set for early stopping and checkpoint selection. Test {metric} was {your score} vs {baseline}. The main errors were {confusion pattern}, which I found by {error analysis}. I exported the model with {export format} and served it via {serving}, measuring {latency} per request on {hardware}. Next I'd {improvement}."*

**Likely follow-ups:**
1. Why a pretrained model (or why not)? What did you freeze and why?
2. How did you know when to stop training? Show me an overfitting sign in your curves.
3. What does the confusion matrix say — which classes get mixed up and why?
4. Which augmentations did you use, and could any of them change the label?
5. What would you do if the model had to run on a phone or a CPU-only server?
6. How did you make the run reproducible?

<details><summary>What good answers cover</summary>

- **Transfer learning:** little data → pretrained backbone, freeze early layers first, then fine-tune with a lower learning rate.
- **Stopping:** validation loss rising while training loss falls = overfitting; early stopping + keeping the best checkpoint; the test set only once at the end.
- **Error analysis:** looking at actual misclassified examples, not only aggregate accuracy; per-class metrics for imbalance.
- **Augmentation validity:** e.g. horizontal flips are wrong for digits or text-like images where orientation carries meaning.
- **Deployment constraints:** smaller architecture, quantization, ONNX/TorchScript-style export, batch size and latency measured on the target hardware.
- **Reproducibility:** seeds, pinned versions, logged configs and metrics (MLflow or W&B from module 09).

</details>

### 📚 Capstone 03 — LLM RAG Assistant Project

**Pitch skeleton:** *"{Users} needed answers from {documents}, and {pain today}. I built a retrieval-augmented assistant: documents are chunked {chunking strategy}, embedded with {embedding model}, retrieved with {retrieval — e.g. hybrid keyword + vector}, and the model answers with citations. I wrote an evaluation set of {n} questions and measured retrieval ({recall@k / MRR}) separately from answer quality ({faithfulness / correctness method}). Compared with {baseline}, {your metric change}. The biggest failure mode was {failure}, which I addressed with {fix}. It runs on {model / hosting} at {latency or cost per query}. Next I'd {improvement}."*

**Likely follow-ups:**
1. How did you choose chunk size and overlap? What happened when you changed it?
2. How do you know an answer is grounded in the retrieved text rather than invented?
3. How did you build the evaluation set, and how big is it? Is it biased toward easy questions?
4. Why hybrid retrieval (or why not)? When does keyword search beat embeddings?
5. What does a query cost, and how would you cut latency and cost by 5×?
6. How do you handle prompt injection hidden inside a document?

<details><summary>What good answers cover</summary>

- **Chunking:** a measured comparison on your eval set, not a default copied from a tutorial; the trade-off between precise citations (small chunks) and complete context (large chunks).
- **Grounding:** citations required, an "I don't know" path when retrieval scores are low, and a faithfulness check (human spot checks or an LLM judge validated against human labels).
- **Eval set:** written *before* tuning, covering hard cases (multi-hop, no-answer, near-duplicate documents); retrieval and generation evaluated separately so you know which half failed.
- **Hybrid search:** keywords win on exact IDs, error codes and rare names; embeddings win on paraphrases; rank fusion combines them.
- **Cost & latency:** caching, smaller models for easy queries, fewer/shorter retrieved chunks, streaming for perceived latency.
- **Injection:** treat retrieved text as data, not instructions; restrict what the model can do; test with adversarial documents.

</details>

### 🤖 Capstone 04 — AI Agent Project

**Pitch skeleton:** *"The task was {multi-step task} for {users}, which previously took {manual effort}. I built an agent that plans and calls {n tools — list them} through {framework / plain tool-calling loop}. Guardrails: {input validation, allowed actions, step limits, human approval for risky actions}. I evaluated it on {n} scripted tasks, measuring {task success rate, steps, cost, failure categories}; versus {baseline — e.g. a single prompt or fixed workflow}, {your result}. The most common failure was {failure}, fixed by {fix}. Every run is traced with {tracing}, so I can replay failures. Next I'd {improvement}."*

**Likely follow-ups:**
1. Why an agent at all — would a fixed workflow have been simpler and more reliable?
2. What stops the agent from looping forever or taking a harmful action?
3. How do you evaluate something non-deterministic? How many runs per task?
4. What happens when a tool fails, times out, or returns malformed output?
5. How do you debug a bad run?
6. What does one task cost in tokens and time, and how would you reduce it?

<details><summary>What good answers cover</summary>

- **Agent vs workflow:** use a fixed pipeline when the steps are known; an agent only when the path genuinely depends on intermediate results. Saying "I started with a workflow and only added autonomy where it helped" is a strong answer.
- **Safety:** max steps/time/budget, an allow-list of tools, schema-validated arguments, read-only by default, human confirmation for irreversible actions.
- **Evaluation:** a fixed task suite with programmatic success checks, several runs per task to report a success *rate*, failure categories counted, not anecdotes.
- **Tool errors:** retries with backoff for transient errors, structured error messages returned to the model, graceful give-up.
- **Debugging:** traces of every model call and tool call with inputs/outputs; replay a failing trace after a fix.
- **Cost:** fewer steps via better tool descriptions, smaller models for simple sub-steps, caching tool results.

</details>

> 💡 **Interview angle:** A course capstone is a perfectly good interview project **if you go one step beyond the tutorial** — a different dataset, an extra baseline, an honest error analysis, or a deployment detail you debugged yourself. That "one step beyond" is what gives you original things to say in the deep dive, and it is the part to lead with.

## 5. Common Behavioral Questions 🟡

Thirty questions that come up again and again in ML/AI engineer loops, grouped by what they test. For each: answer **out loud** with a 2-minute limit, then open the hidden answer.

> ⚠️ **Every "30-second answer" is a labelled 📝 *example* about a made-up situation.** It shows the *shape* of a strong answer — specific, "I"-focused, with a result and a reflection. Do not memorise or borrow it: follow-ups will expose a story that isn't yours, and presenting someone else's experience as your own is dishonest. Use your story bank from section 2.

### 👋 Openers

**Q6. "Tell me about yourself."**

<details><summary>Show answer</summary>

- **30-second answer:** 📝 *Example:* "I'm a data analyst moving into ML engineering. Over the last two years I automated our reporting pipelines in Python and SQL, which is where I got hooked on building things that run reliably. In the past year I've built an end-to-end churn model served behind an API and a document Q&A assistant with an evaluation set I wrote myself. I'm looking for a role where I can own models from data to production, which is why this position caught my eye."
- **Go deeper:** Use **present → past → future** in 60–90 seconds: who you are now, 2–3 relevant highlights, why this role. It's a trailer, not your life story — end on a hook they'll want to ask about.
- **❌ Common wrong answer:** Reciting the résumé chronologically from school, or talking about hobbies for two minutes.

</details>

**Q7. "Why do you want to work here / on this team?"**

<details><summary>Show answer</summary>

- **30-second answer:** 📝 *Example:* "Three reasons. Your team ships search ranking to millions of users, and retrieval is what I've enjoyed most in my projects. Your engineering blog post on evaluating ranking changes offline before A/B tests matches how I like to work. And the role mixes modelling with production work, which is exactly the growth I'm after."
- **Go deeper:** Be specific to *this* company: product, technical challenges, a talk or blog post, the team's stack or problem domain. Connect each reason to your experience.
- **❌ Common wrong answer:** "It's a great company with a great culture" — true of anywhere, so it says nothing.

</details>

**Q8. "Why are you moving into ML engineering?" (career switchers)**

<details><summary>Show answer</summary>

- **30-second answer:** 📝 *Example:* "In my operations job I kept building small forecasting spreadsheets, then Python scripts, and I noticed I enjoyed that more than anything else in the role. I started taking it seriously: I completed a structured course, built two end-to-end projects, and automated a scheduling report at work that the team still uses. The switch is a continuation of what I was already doing."
- **Go deeper:** Frame your previous career as an **asset** (domain knowledge, stakeholder skills, operational rigour) and show evidence of commitment (projects, not just certificates).
- **❌ Common wrong answer:** "ML is where the money is" or "AI is the future" with no personal story.

</details>

### 💥 Failure & mistakes

**Q9. "Tell me about a time you failed."**

<details><summary>Show answer</summary>

- **30-second answer:** 📝 *Example (CARL):* "**C:** For a course project I promised a demo of a sentiment model in two weeks. **A:** I spent ten days tuning a transformer before looking closely at the labels, then discovered a third were noisy duplicates. I told my team straight away, re-scoped the demo to a cleaned subset, and wrote a short label-audit script. **R:** The demo was a day late and on less data. **L:** Now I spend the first day on data quality and a simple baseline before any tuning — and I've kept that habit on every project since."
- **Go deeper:** Choose a **real** failure with real consequences, own your part without self-flagellation, and spend at least a third of the answer on what changed afterwards. Interviewers probe: "What would you do differently?", "How did others react?"
- **❌ Common wrong answer:** A disguised success ("I cared too much about quality") or a failure that was entirely someone else's fault.

</details>

**Q10. "Tell me about a mistake that affected users or production."**

<details><summary>Show answer</summary>

- **30-second answer:** 📝 *Example:* "I deployed a retrained recommendation model without checking that a feature's units had changed from seconds to milliseconds. Click-through dropped noticeably the next morning. I rolled back to the previous model within the hour, confirmed the cause by comparing feature distributions, and added a pre-deployment check that compares feature statistics with the training data. The same check blocked a similar issue two months later."
- **Go deeper:** Show the **incident sequence**: detect → mitigate (rollback first) → find root cause → prevent recurrence. Mention communicating to affected people and a blameless write-up if your team does them.
- **❌ Common wrong answer:** Minimising ("it wasn't a big deal") or spending the answer on who else should have caught it.

</details>

**Q11. "What's your biggest weakness?" / "What are you working on improving?"**

<details><summary>Show answer</summary>

- **30-second answer:** 📝 *Example:* "I tend to go deep on the modelling side and under-invest in writing things down. On my last project a teammate couldn't reproduce my results because the preprocessing lived in my head. Since then I write a one-page design note before starting and keep configs in version control — it's improving, and my last project had a README someone else used to rerun it end to end."
- **Go deeper:** A real, non-fatal-for-this-role weakness + concrete evidence you're working on it. Weaknesses in skills you can learn (public speaking, documentation, a technology) are safer than core-job gaps.
- **❌ Common wrong answer:** "I'm a perfectionist" / "I work too hard" — interviewers hear it constantly and read it as evasive.

</details>

### 🤝 Conflict & collaboration

**Q12. "Tell me about a conflict with a teammate."**

<details><summary>Show answer</summary>

- **30-second answer:** 📝 *Example:* "A teammate wanted to report overall accuracy for our fraud model; I thought it hid the fact that we missed most fraud. Rather than argue in the group chat, I asked for 20 minutes, showed a confusion matrix, and explained what each error cost the business. We agreed to report recall at a fixed false-positive rate, plus accuracy for continuity. The review went smoothly and we still work together well."
- **Go deeper:** Show **curiosity about their view**, a move from opinions to evidence, a decision, and a preserved relationship. Say what you'd do if the evidence had gone the other way.
- **❌ Common wrong answer:** "I've never had a conflict" (not credible) or a story where the other person is simply wrong and you "won".

</details>

**Q13. "Tell me about a time you disagreed with your manager."**

<details><summary>Show answer</summary>

- **30-second answer:** 📝 *Example:* "My manager wanted to launch a model on Friday. I'd noticed validation performance dropped for new customers, who were a small but growing segment. I brought a one-slide breakdown by segment and suggested launching for existing customers only while I fixed the gap. He agreed; the full launch happened two weeks later with the segment gap closed."
- **Go deeper:** Disagree **privately, with data, offering options** — then *commit* once a decision is made, even if it isn't yours ("disagree and commit"). Mention how you'd escalate if a real risk (legal, safety) were ignored.
- **❌ Common wrong answer:** Going around the manager, or "I just did what I was told" when you had real concerns.

</details>

**Q14. "Tell me about a time a teammate wasn't pulling their weight."**

<details><summary>Show answer</summary>

- **30-second answer:** 📝 *Example:* "In a four-person capstone, one teammate hadn't pushed code for two weeks. I messaged him privately to ask how things were going; it turned out he was stuck on the Docker setup and embarrassed to say. We paired for an hour, split his task into smaller pieces with check-ins, and he delivered the API part in time."
- **Go deeper:** Start from **curiosity, not judgement** — assume there's a reason. Talk privately first, make the work visible (smaller tasks, check-ins), and escalate to a lead only if that fails.
- **❌ Common wrong answer:** "I just did their work myself" (burnout, no fix) or complaining publicly.

</details>

**Q15. "Tell me about a time you received critical feedback."**

<details><summary>Show answer</summary>

- **30-second answer:** 📝 *Example:* "In a code review, a senior engineer said my feature pipeline was untestable — everything was one 300-line function. My first reaction was defensive, but I asked her to show me how she'd structure it. I split it into pure functions with unit tests, which caught a bug in date handling. I now ask for design feedback before writing big pieces."
- **Go deeper:** Show the full loop: hear it → ask clarifying questions → act → see a result → change a habit. Admitting an initial emotional reaction makes it believable.
- **❌ Common wrong answer:** Feedback that was secretly a compliment, or explaining why the reviewer was wrong.

</details>

### ⏰ Deadlines & prioritisation

**Q16. "Tell me about a time you missed a deadline."**

<details><summary>Show answer</summary>

- **30-second answer:** 📝 *Example:* "I estimated three days to add a new data source to our training set; it took eight because the source had no documentation and inconsistent IDs. My mistake was telling my lead on day four instead of day one, when I first saw the mess. I now flag risks as soon as I see them with a revised estimate and options — ship without the source, or wait."
- **Go deeper:** Interviewers care less about the miss than about **when you communicated** and what you learned about estimating. Mention breaking work down and time-boxing unknowns.
- **❌ Common wrong answer:** "I've never missed a deadline" or blaming the data team entirely.

</details>

**Q17. "How do you prioritise when you have too many things to do?"**

<details><summary>Show answer</summary>

- **30-second answer:** 📝 *Example:* "Last quarter I had a model retrain, two stakeholder analysis requests and an on-call rotation in the same week. I listed everything with impact and urgency, asked my manager to confirm the top two, and told the analysts honestly that their requests would land next week, offering a quick partial answer for one. The retrain shipped on time and neither analyst was surprised."
- **Go deeper:** Show a method (impact × urgency, what's blocking others, what's reversible) and — crucially — **communicating trade-offs** instead of silently dropping things.
- **❌ Common wrong answer:** "I just work longer hours" — it doesn't scale and signals poor judgement.

</details>

**Q18. "Tell me about a time you had to trade quality for speed."**

<details><summary>Show answer</summary>

- **30-second answer:** 📝 *Example:* "A sales team needed lead scores for a campaign in one week. Instead of the planned boosted model with full tuning, I shipped a logistic regression with five well-understood features and a documented list of known limitations, then built the better model afterwards. The quick model was good enough for the campaign, and the limitations note meant nobody over-trusted it."
- **Go deeper:** Separate **reversible shortcuts** (a simpler model, less tuning) from **irreversible ones** (skipping evaluation, leaking test data, privacy shortcuts) — take the first kind, refuse the second, and write down the debt.
- **❌ Common wrong answer:** "I never compromise on quality" (unrealistic) or cutting corners on evaluation without saying so.

</details>

### 🌫️ Ambiguity & incomplete data

**Q19. "Tell me about a time the requirements were ambiguous."**

<details><summary>Show answer</summary>

- **30-second answer:** 📝 *Example:* "A product manager asked for 'AI to make support faster'. I booked 30 minutes with two support agents, watched them work, and found most time went to searching old tickets. I proposed a narrow first version — similar-ticket search — with one success measure, time to first response, and confirmed it with the PM before building. The pilot was scoped in a week instead of drifting for a month."
- **Go deeper:** The pattern is **clarify the goal → talk to users → write down assumptions → ship the smallest useful thing → measure**. Mention what you'd do if stakeholders disagreed on the goal.
- **❌ Common wrong answer:** Waiting for someone to write perfect requirements, or building a big system on your own guess.

</details>

**Q20. "Tell me about a decision you made with incomplete data."**

<details><summary>Show answer</summary>

- **30-second answer:** 📝 *Example:* "We had to choose between two embedding models for search before we had any user queries. I built a 40-question test set from real help-centre articles, measured recall@5 for both, and picked the cheaper one because the gap was within noise on that small set. I documented it as a reversible decision and set a reminder to re-run the comparison once we had a month of real queries — which confirmed the choice."
- **Go deeper:** Show that you (1) gathered the *cheapest* useful evidence, (2) judged whether the decision was reversible, (3) stated your confidence, and (4) planned a check. Reversible decisions should be made quickly.
- **❌ Common wrong answer:** "I waited until we had all the data" — or gut feel with no plan to verify.

</details>

**Q21. "Tell me about a time your model's results were disappointing. What did you do?"**

<details><summary>Show answer</summary>

- **30-second answer:** 📝 *Example:* "My first demand-forecast model barely beat a 'same as last week' baseline. Instead of trying bigger models, I broke errors down by product and found almost all of the error came from promotional weeks, which the data didn't mark. I worked with the marketing analyst to add a promotion calendar feature, which gave a clear improvement on those weeks, and I reported honestly that other weeks were already near the baseline."
- **Go deeper:** Show **error analysis before model changes**, respect for a strong simple baseline, and honest reporting — sometimes the right answer is "the model isn't worth deploying".
- **❌ Common wrong answer:** "I tried a bigger neural network until the number went up" — especially if tuned on the test set.

</details>

### 🗣️ Influence & pushing back

**Q22. "Tell me about a time you pushed back on a stakeholder."**

<details><summary>Show answer</summary>

- **30-second answer:** 📝 *Example:* "A marketing lead wanted the churn model to go live in all regions next week. The model had only been validated on one region's data. I explained the risk in her terms — wasted discount budget on the wrong customers — and offered a two-week pilot in two regions with a holdout group. She agreed; the pilot showed one region needed retraining, which we fixed before the full launch."
- **Go deeper:** Push back on the **plan, not the person**: acknowledge the goal, name the risk in their currency (money, customers, time), and always bring an alternative that still moves them forward.
- **❌ Common wrong answer:** "I told them it wasn't possible" with no alternative, or quietly agreeing and shipping something you believed was risky.

</details>

**Q23. "Tell me about a time you influenced people without authority."**

<details><summary>Show answer</summary>

- **30-second answer:** 📝 *Example:* "Every team on our floor evaluated models differently, so results weren't comparable. I wasn't anyone's lead, so I built a small shared evaluation script for my own team, wrote a one-page guide, and offered to help one other team adopt it. Once two teams were using it, a third asked to join, and it became the default in the team's model-review template."
- **Go deeper:** Influence comes from **making the right thing easy** (tools, templates), showing results, and finding early allies — not from arguing harder.
- **❌ Common wrong answer:** "I escalated to their manager" as the first move.

</details>

**Q24. "A stakeholder insists on using the newest, biggest AI model for a simple problem. How do you handle it?"**

<details><summary>Show answer</summary>

- **30-second answer:** 📝 *Example:* "A director wanted a large LLM to route support emails into five categories. I agreed to test it fairly: I labelled 300 emails and compared the LLM with a small fine-tuned classifier on accuracy, latency and cost per thousand emails. The small model was nearly as accurate at a fraction of the cost and latency, so we used it and kept the LLM for the hard 'other' cases."
- **Go deeper:** Don't dismiss the idea — **turn it into an experiment** with agreed metrics (quality, latency, cost, maintenance). Hybrid answers (cheap model first, big model for hard cases) often win.
- **❌ Common wrong answer:** "LLMs are overhyped" (dismissive) or "sure" with no evaluation.

</details>

### 🧑‍🏫 Communicating with non-technical people

**Q25. "Explain one of your models to someone non-technical." / "Tell me about a time you did."**

<details><summary>Show answer</summary>

- **30-second answer:** 📝 *Example:* "I explained our churn model to the retention team like a weather forecast: it doesn't know who *will* cancel, it gives each customer a chance of cancelling, based on patterns from past customers — like fewer logins and more support tickets. I showed that if they called the top 10% riskiest customers, they'd reach far more future cancellers than calling at random, using our validation numbers. They started using the ranked list the next week."
- **Go deeper:** Lead with **what it's for and how to use it**, use an everyday analogy, show one concrete example, state what it can't do, then check understanding ("How would you use this list on Monday?").
- **❌ Common wrong answer:** Starting with the algorithm ("It's a gradient-boosted ensemble of decision trees…").

</details>

**Q26. "How would you explain to a business stakeholder why the model is sometimes wrong?"**

<details><summary>Show answer</summary>

- **30-second answer:** 📝 *Example:* "When a manager was upset that the fraud model flagged a loyal customer, I explained that every threshold trades two kinds of mistakes: flag more and you annoy good customers, flag less and you miss fraud. I showed both error counts at three thresholds on last month's data and asked which trade-off the business preferred. They chose a threshold and added a fast manual-review path for high-value customers."
- **Go deeper:** Make uncertainty **a business decision they own**: frame errors as costs, show the trade-off curve in plain numbers, and design a process around the mistakes (review queues, appeals).
- **❌ Common wrong answer:** "Models aren't perfect" with nothing actionable.

</details>

**Q27. "Tell me about a time you explained a technical trade-off to leadership."**

<details><summary>Show answer</summary>

- **30-second answer:** 📝 *Example:* "Leadership asked why our assistant needed two more months. I wrote a one-page memo with three options — launch now with a known 1-in-10 wrong-citation rate on our test set, launch to internal users only, or wait for the fix — each with its risk, cost and date. They chose the internal launch, which also gave us real queries to evaluate with."
- **Go deeper:** Leaders need **options, risks, and a recommendation** — not architecture. Put the decision in the first sentence and the details in an appendix.
- **❌ Common wrong answer:** Explaining the technical details in depth and leaving the decision implicit.

</details>

### ⚖️ Ethics, fairness & responsibility

**Q28. "Tell me about a time you raised an ethics or fairness concern."**

<details><summary>Show answer</summary>

- **30-second answer:** 📝 *Example:* "While building a résumé-screening prototype for a class, I noticed the historical 'hired' labels came from a period with very few women hired, so the model was learning past decisions, not ability. I measured selection rates by gender on the validation set, showed the gap to my project group, and we changed the project to ranking by skills-match with the sensitive attributes and obvious proxies removed, plus a documented fairness check. We also wrote the limitations into the README."
- **Go deeper:** Show that you (1) noticed a concrete risk, (2) **measured it** rather than asserting it, (3) raised it early with options, and (4) documented limitations (e.g. a model card). Acknowledge that removing sensitive columns alone doesn't remove bias, because proxies remain.
- **❌ Common wrong answer:** "Ethics isn't really my area — that's for the legal team," or claiming a model is unbiased because it doesn't use the protected attribute.

</details>

**Q29. "What would you do if you were asked to report a result you didn't believe was valid?"**

<details><summary>Show answer</summary>

- **30-second answer:** 📝 *Example:* "A teammate wanted to present a model's score on data that had also been used for tuning. I explained that the number would be optimistic and proposed a quick re-run on a held-out slice before the meeting. The honest number was lower; we presented it with the reason, and the review focused on fixing the gap instead of discovering it later in production."
- **Go deeper:** Separate honest mistakes (fix quietly and quickly) from pressure to mislead (escalate calmly with evidence). Offer a path that still meets the deadline: a caveated number, or a smaller valid result.
- **❌ Common wrong answer:** "I'd present it anyway — it's their call" or publicly accusing the teammate.

</details>

**Q30. "How do you handle sensitive or personal data in a project?"**

<details><summary>Show answer</summary>

- **30-second answer:** 📝 *Example:* "In a health-related side project, I only used a public, de-identified dataset, kept raw files out of Git, logged aggregate metrics only, and checked that no example rows appeared in notebook outputs or in the model's prompts. When a teammate suggested scraping patient-forum posts, I pointed out consent and terms-of-service problems and we didn't."
- **Go deeper:** Mention data minimisation (collect only what you need), access control, not sending personal data to third-party APIs without approval, retention limits, and following the company's legal/privacy review. You don't need to be a lawyer — you need to know when to ask one.
- **❌ Common wrong answer:** "It's anonymised, so anything is fine" — re-identification from combinations of fields is a known risk.

</details>

### 📚 Learning & growth

**Q31. "Tell me about a time you had to learn a new technology quickly."**

<details><summary>Show answer</summary>

- **30-second answer:** 📝 *Example:* "I joined a project that used Spark with two weeks until a data deadline, and I'd only used pandas. I did the official quick-start on day one, then rewrote one small existing job myself and asked a teammate to review it, instead of reading documentation for a week. By the deadline I'd built the feature aggregation job, and my review notes became the onboarding doc for the next person."
- **Go deeper:** Show a **learning strategy**: official docs/tutorial → a small real task → feedback from someone experienced → teach it back. Mention what you deliberately *didn't* learn yet.
- **❌ Common wrong answer:** "I'm a fast learner" with no specific example, or a list of courses watched.

</details>

**Q32. "How do you keep up with a field that changes as fast as AI?"**

<details><summary>Show answer</summary>

- **30-second answer:** 📝 *Example:* "I follow a few primary sources — release notes and papers from the tools I use — rather than trying to read everything. Once a month I pick one idea and actually try it on a small problem; last month I compared two reranking approaches on my RAG test set. Trying things on my own eval set tells me more than headlines do."
- **Go deeper:** Filtering matters more than volume: primary sources over hype, and evaluation on **your** data before adopting anything. Mention how you share what you learn (notes, team demos).
- **❌ Common wrong answer:** Listing ten newsletters and podcasts with no evidence of applying anything.

</details>

### 🏆 Impact & ownership

**Q33. "What's the most impactful thing you've done?"**

<details><summary>Show answer</summary>

- **30-second answer:** 📝 *Example:* "At my last job the monthly inventory report took an analyst two full days to assemble by hand and often had errors. I automated it with a tested Python pipeline and a validation step that flags missing stores. It now runs in under an hour, errors reported by store managers dropped to almost none, and the analyst uses those two days for actual analysis."
- **Go deeper:** Impact = **who benefited and how much**, ideally with a number and a comparison. Be ready to explain how the number was measured and what part was yours. Mid-sized, well-measured impact beats vague huge claims.
- **❌ Common wrong answer:** "I built a state-of-the-art model" with no user, no comparison, and no measured effect.

</details>

**Q34. "Tell me about a time you went beyond your role."**

<details><summary>Show answer</summary>

- **30-second answer:** 📝 *Example:* "Our team's model training jobs failed silently about once a week and nobody owned the alerting. It wasn't in my tasks, but I spent a Friday adding failure alerts and a short runbook, then asked my lead to put it on the team's rotation. Silent failures stopped, and the runbook got used twice in the next month by other people."
- **Go deeper:** Ownership means **noticing an unowned problem, fixing it or finding it an owner, and making the fix last** — without neglecting your actual responsibilities (say how you balanced them).
- **❌ Common wrong answer:** A story about working all weekend as the main point, or taking over someone else's area without telling them.

</details>

**Q35. "Which project are you proudest of, and what would you do differently?"**

<details><summary>Show answer</summary>

- **30-second answer:** 📝 *Example:* "My document Q&A assistant, because I built the evaluation set before tuning anything, so every change was measured instead of eyeballed. If I did it again, I'd collect real user questions earlier: my hand-written test questions turned out to be easier than what people actually ask."
- **Go deeper:** Pick the project you can deep-dive for 45 minutes (section 3). The "differently" half is the real test — a specific, technical, self-aware answer shows growth.
- **❌ Common wrong answer:** "Nothing, it went perfectly," or choosing a project you can't explain in depth.

</details>

## 6. Writing Résumé Bullets for ML Projects 🟢

A résumé bullet is a **one-line trailer** for a story you are ready to tell in full. Recruiters skim; engineers who read it will pick a bullet and ask you about it. A widely used shape is:

> **Action verb** + **what you built, with the method** + **measurable result** (compared with something)

| Part | ❌ Weak | ✅ Better |
|---|---|---|
| Action | "Responsible for models" | "Built", "Cut", "Designed", "Automated", "Diagnosed" |
| Method | "using AI" | "gradient boosting in a leak-free scikit-learn pipeline" |
| Result | "improved performance" | "ROC-AUC from {baseline} to {yours} on a held-out test set" |

### Common mistakes

1. **Tools without outcomes** — "Python, PyTorch, Docker, AWS" belongs in a skills line, not a project bullet.
2. **Weak openers** — "Responsible for", "Worked on", "Helped with" hide what *you* did.
3. **Unverifiable claims** — "significantly improved", "state-of-the-art", "various models".
4. **Inflated or context-free metrics** — "99% accuracy" on a dataset that is 99% one class; "increased revenue 30%" when you cannot explain how it was measured or what share was yours.
5. **Walls of text** — more than ~2 lines per bullet gets skimmed past.

**Honesty test for every bullet:** *Could I defend this number for 10 minutes, including how it was measured, what it was compared against, and what I personally did?* If not, soften it to what you can defend ("on a 60-question test set I wrote") or remove it.

### The résumé-bullet checker

A small, transparent rule-based checker. Each rule is a few lines you can read — no model, no magic. It checks **shape, not truth**: it cannot tell whether a number is real. That part is on you.

| Rule | Flags a bullet when… |
|---|---|
| `no_metric` | there is no quantity — years (`2024`), names with digits (`GPT-4`, `YOLOv8`, `F1`, `p95`) and versions after a tool (`Python 3.12`) don't count |
| `weak_opener` | it starts with a passive phrase like "Responsible for" or "Worked on" |
| `tool_list` | it names ≥3 tools and either has no metric or is at least half tool names |
| `too_long` | it has more than 30 words |
| `check_claim` | it claims ≥99% accuracy/precision/recall/F1 — not necessarily wrong, but be ready to explain class balance and leakage checks |

In [10]:
WEAK_OPENERS = ("responsible for", "worked on", "helped with", "helped", "assisted with", "assisted",
                "involved in", "participated in", "tasked with", "duties included", "in charge of", "exposure to")
TOOLS = {"python", "pytorch", "tensorflow", "keras", "scikit-learn", "sklearn", "pandas", "numpy", "polars", "sql",
         "postgres", "docker", "kubernetes", "aws", "gcp", "azure", "fastapi", "flask", "mlflow", "airflow", "dvc",
         "spark", "pyspark", "langchain", "langgraph", "llamaindex", "faiss", "xgboost", "lightgbm", "catboost", "git",
         "transformers", "huggingface", "openai", "opencv", "jupyter", "streamlit", "redis", "linux", "tableau", "excel"}
STOPWORDS = {"a", "an", "the", "and", "or", "of", "to", "in", "on", "for", "with", "using", "used", "via", "by",
             "from", "into", "at", "as", "&", "+"}
EDGE_PUNCT = ",.;:()[]\"'!?"


def clean_tokens(text):
    return [tok.strip(EDGE_PUNCT) for tok in text.split() if tok.strip(EDGE_PUNCT)]


def metric_tokens(bullet):
    """Tokens that look like real quantities (not years, versions, or model names)."""
    tokens = clean_tokens(bullet)
    found = []
    for i, tok in enumerate(tokens):
        if not any(ch.isdigit() for ch in tok):
            continue
        if re.fullmatch(r"(19|20)\d\d", tok):                      # a year
            continue
        if tok[0].isalpha() and not tok.endswith("%"):             # GPT-4, YOLOv8, F1, p95, recall@5 (but keep top-10%)
            continue
        if i > 0 and tokens[i - 1].lower() in TOOLS:               # Python 3.12, pandas 3.0
            continue
        found.append(tok)
    return found


def tool_mentions(bullet):
    return [tok for tok in clean_tokens(bullet) if tok.lower() in TOOLS]


def suspicious_claim(bullet):
    low = bullet.lower()
    for m in re.finditer(r"(\d+(?:\.\d+)?)\s?%", low):
        window = low[max(0, m.start() - 25): m.end() + 25]
        if float(m.group(1)) >= 99 and re.search(r"accura|precision|recall|f1", window):
            return True
    return False


def review_bullet(bullet, max_words=30):
    """Return a list of rule names the bullet breaks ([] means it passes every rule)."""
    text = bullet.strip().lstrip("-•*▪ ").strip()
    issues = []
    metrics = metric_tokens(text)
    if not metrics:
        issues.append("no_metric")
    if text.lower().startswith(WEAK_OPENERS):
        issues.append("weak_opener")
    tools = tool_mentions(text)
    content = [t for t in clean_tokens(text) if t.lower() not in STOPWORDS]
    if len(tools) >= 3 and (not metrics or len(tools) / len(content) >= 0.5):
        issues.append("tool_list")
    if len(text.split()) > max_words:
        issues.append("too_long")
    if suspicious_claim(text):
        issues.append("check_claim")
    return issues


# Illustrative bullets — the "strong" ones use made-up numbers purely to show the format
sample_bullets = [
    "Responsible for building machine learning models for the data science team.",
    "Python, PyTorch, Docker, FastAPI, MLflow, AWS",
    "Worked on a churn prediction project using Python 3.12, pandas and scikit-learn in 2024.",
    "Achieved 99.8% accuracy on credit-card fraud detection with XGBoost.",
    "Built a churn model in scikit-learn that raised ROC-AUC from 0.71 (logistic baseline) to 0.84 on a held-out test set.",
    "Cut p95 API latency from 800 ms to 120 ms by batching requests and caching embeddings in Redis.",
    "Designed and implemented an end-to-end machine learning pipeline including data ingestion, validation, feature "
    "engineering, hyperparameter tuning with Optuna, experiment tracking, containerisation, monitoring and deployment for the "
    "analytics team, reducing manual work by 10 hours per week.",
]
pd.DataFrame({"bullet": sample_bullets,
              "issues": [", ".join(review_bullet(b)) or "✅ passes" for b in sample_bullets],
              "quantities found": [metric_tokens(b) for b in sample_bullets]})

,bullet,issues,quantities found
0,Responsible for building machine learning models for the data science team.,"no_metric, weak_opener",[]
1,"Python, PyTorch, Docker, FastAPI, MLflow, AWS","no_metric, tool_list",[]
2,"Worked on a churn prediction project using Python 3.12, pandas and scikit-le...","no_metric, weak_opener, tool_list",[]
3,Achieved 99.8% accuracy on credit-card fraud detection with XGBoost.,check_claim,[99.8%]
4,Built a churn model in scikit-learn that raised ROC-AUC from 0.71 (logistic ...,✅ passes,"[0.71, 0.84]"
5,Cut p95 API latency from 800 ms to 120 ms by batching requests and caching e...,✅ passes,"[800, 120]"
6,Designed and implemented an end-to-end machine learning pipeline including d...,too_long,[10]


Reading the table: bullet 3 mentions `3.12` and `2024`, yet it is still flagged `no_metric` — neither is a result. Bullet 4 has a number but trips `check_claim`: fraud datasets are usually heavily imbalanced, so an interviewer's first question will be "what fraction of transactions were fraud?". The last bullet has a real result buried after 30+ words.

**Limits (say these if asked about the tool):** a rule-based checker misses synonyms it hasn't been told about, can't judge whether a verb is strong, and can't verify truth. It is a linter, like `flake8` for prose — useful for catching the obvious, useless as a judge of quality.

### ✍️ Your Turn — rewrite a weak bullet

Here is an **illustrative fact sheet** for a project (pretend these are *your* measured results):
- Built a document Q&A assistant over an internal policy manual.
- Retrieval: hybrid BM25 + embeddings; recall@5 went from **0.62** (embeddings only) to **0.81** on a **60-question** test set you wrote.
- Answers include citations; served via FastAPI.

The weak bullet is: *"Worked on a RAG chatbot using LangChain, FAISS and OpenAI."* Write a bullet using **only the facts above** so that `review_bullet(your_bullet)` returns `[]`.

In [11]:
weak_bullet = "Worked on a RAG chatbot using LangChain, FAISS and OpenAI."
print("weak bullet issues:", review_bullet(weak_bullet))

your_bullet = None  # TODO: a string
check("rewritten bullet passes the checker", None if your_bullet is None else review_bullet(your_bullet), [],
      hint="Start with an action verb, include the before → after metric and the test-set size, stay ≤ 30 words.")

weak bullet issues: ['no_metric', 'weak_opener', 'tool_list']
⏳ rewritten bullet passes the checker: not attempted yet — replace None with your answer.


<details><summary>💡 Show solution</summary>

```python
weak_bullet = "Worked on a RAG chatbot using LangChain, FAISS and OpenAI."
print("weak bullet issues:", review_bullet(weak_bullet))

your_bullet = ("Built a document Q&A assistant with hybrid BM25 + embedding retrieval, raising recall@5 from 0.62 "
               "to 0.81 on a 60-question test set; served cited answers via FastAPI.")
check("rewritten bullet passes the checker", review_bullet(your_bullet), [])
print(len(your_bullet.split()), "words | quantities:", metric_tokens(your_bullet))
```

Notice what makes it defensible: the comparison ("from 0.62"), the evaluation size ("60-question test set") and the fact that you wrote the test set — all things you can explain for ten minutes.
</details>

### ✍️ Your Turn — implement a new checker rule 🟡

Implement `find_vague_words(bullet)`: return the **sorted, de-duplicated** list of words from `VAGUE_WORDS` that appear in the bullet as whole words, ignoring case and surrounding punctuation (so `"etc."` counts as `etc`).

In [12]:
VAGUE_WORDS = {"various", "significantly", "greatly", "several", "many", "numerous", "robust",
               "cutting-edge", "state-of-the-art", "etc", "seamlessly", "synergy"}


def find_vague_words(bullet):
    return None  # TODO


vague_tests = [
    "Significantly improved various models using cutting-edge techniques, etc.",
    "Reduced training time from 6 hours to 40 minutes by caching features.",
    "Built ROBUST pipelines for many teams, many times.",
]
got = None if find_vague_words(vague_tests[0]) is None else [find_vague_words(b) for b in vague_tests]
check("find_vague_words", got, [["cutting-edge", "etc", "significantly", "various"], [], ["many", "robust"]],
      hint="Split on whitespace, strip string.punctuation from each token, lowercase, then intersect with VAGUE_WORDS.")

⏳ find_vague_words: not attempted yet — replace None with your answer.


<details><summary>💡 Show solution</summary>

```python
VAGUE_WORDS = {"various", "significantly", "greatly", "several", "many", "numerous", "robust",
               "cutting-edge", "state-of-the-art", "etc", "seamlessly", "synergy"}


def find_vague_words(bullet):
    tokens = {tok.strip(string.punctuation).lower() for tok in bullet.split()}
    return sorted(tokens & VAGUE_WORDS)


vague_tests = [
    "Significantly improved various models using cutting-edge techniques, etc.",
    "Reduced training time from 6 hours to 40 minutes by caching features.",
    "Built ROBUST pipelines for many teams, many times.",
]
got = [find_vague_words(b) for b in vague_tests]
check("find_vague_words", got, [["cutting-edge", "etc", "significantly", "various"], [], ["many", "robust"]])
```

`str.strip(string.punctuation)` only removes characters at the **ends**, so the hyphens inside `cutting-edge` survive. Using a set both de-duplicates ("many" twice) and makes the intersection one operation. To add it to the checker, append `"vague"` to `issues` in `review_bullet` when the list is non-empty.
</details>

**Q36. An interviewer points at a bullet and says "tell me more" — and it's a team project where your part was small. What do you do?**

<details><summary>Show answer</summary>

- **30-second answer:** Be precise about ownership immediately: *"This was a four-person project; the model was a teammate's. I owned the data validation and the evaluation harness — let me walk you through those."* Then go deep on your part.
- **Go deeper:** Fix the résumé afterwards so the bullet describes your contribution ("Built the evaluation harness for…"). Precision about ownership reads as integrity, and it steers the conversation to what you can defend.
- **❌ Common wrong answer:** Describing the whole project as if you did it — the next "why did you choose that architecture?" exposes it.

</details>

> 💡 **Interview angle:** Interviewers often open with *"pick any bullet on your résumé and tell me about it"*. Every bullet should therefore map to a story in your bank — if a bullet has no story behind it, cut it.

## 7. Questions to Ask and Follow-Up Etiquette 🟢

"Do you have any questions for us?" is part of the interview. Good questions show you're thinking about doing the job well — and the answers help *you* decide whether you want it.

| Topic | Example questions | What you learn |
|---|---|---|
| **The work** | "What does the path from a notebook experiment to production look like here?" · "How are models evaluated before launch, and who decides they're good enough?" | ML maturity, whether you'll ship |
| **The role** | "What would success look like for this role after six months?" · "What's a problem the team is stuck on right now?" | Expectations, real priorities |
| **The team** | "How are disagreements about technical direction resolved?" · "How do ML engineers and product managers work together?" | Culture, collaboration |
| **Growth** | "How do people learn new things here — reading groups, conference time, mentoring?" · "What does onboarding look like?" | Support for early-career people |
| **Data & ethics** | "How does the team handle model risks like bias or privacy reviews?" | Whether responsibility is real or a slogan |

**Avoid** questions answered on the first page of their website, and leave salary/benefits for the recruiter conversation unless the interviewer brings it up. Tailor to the person: ask an engineer about the stack and on-call, a manager about priorities and growth.

### Follow-up etiquette

- **Thank-you note:** a short email (3–4 sentences) within about a day, ideally mentioning one specific thing you discussed. It rarely changes a decision, but it's courteous and cheap.
- **Waiting:** ask about the timeline at the end of the interview. If it passes, one polite check-in with the recruiter is fine; repeated chasing is not.
- **Rejection:** reply graciously, ask if any feedback is available (many companies have policies against sharing it — don't push), and keep the door open. People who handled rejection well are often contacted again.
- **Confidentiality:** don't post the exact questions you were asked if you agreed not to share them, and never record an interview without explicit consent.
- **Honesty after the fact:** if you realise you said something wrong in the interview (e.g. a wrong formula), a brief correction in the thank-you note is fine and can come across well.

**Q37. The interviewer asks "Do you have any questions for me?" and you've run out of prepared ones. What do you say?**

<details><summary>Show answer</summary>

- **30-second answer:** Ask about something *they* said during the interview: *"You mentioned the team retrains the ranking model weekly — what usually breaks when that happens?"* It shows you were listening and is always available.
- **Go deeper:** Keep one or two universal questions in reserve ("What do you enjoy most about working here?", "What would make someone really successful in their first six months?"). Ending with a clear, warm close ("Thanks — I enjoyed talking about X") matters too.
- **❌ Common wrong answer:** "No, I think you covered everything." It signals low interest.

</details>

> 💡 **Interview angle:** Behavioral questions are often **the same question in disguise** — "a time you disagreed", "a time you pushed back", "a conflict with a stakeholder" can all be answered from one story with a different spotlight. Listen for the competency being tested, then pick the story from your bank that shows it best.

## 🏋️ Timed Drills

Three mock sets. Do one set per sitting, **out loud**, ideally recorded on your phone or with a friend playing interviewer.

**How the 2-minute timer works:**
1. Read the question, then take **up to 10 seconds** to pick a story from your bank (you'd pause like this in a real interview too).
2. Start a **2-minute** timer and answer. Aim to finish the Result and Learning by roughly 1:45.
3. When the timer rings, stop — even mid-sentence. If you were cut off before the Result, that's the most important thing to fix.
4. Have your partner (or yourself, playing back the recording) ask **one follow-up** from the list and answer it in under 1 minute.
5. Score the answer with the rubric below *before* moving on.

### Set A — Hiring-manager round (warm-up)

| # | Question | Follow-up to expect |
|---|---|---|
| A1 | Tell me about yourself. | "Why did you leave / want to leave your current role?" |
| A2 | Tell me about a time you failed. | "What would you do differently today?" |
| A3 | Tell me about a time the requirements were unclear. | "How did you know the first version was good enough?" |
| A4 | Explain one of your projects to me as if I were a product manager. | "How would I know if it's working?" |
| A5 | Why this role? | "What would you want to learn in your first six months?" |

### Set B — Dedicated behavioral round (pressure)

| # | Question | Follow-up to expect |
|---|---|---|
| B1 | Tell me about a conflict with a teammate. | "What did they say exactly? What if they still disagreed?" |
| B2 | Tell me about a time you missed a deadline. | "When did you first know it was at risk?" |
| B3 | Tell me about a time you pushed back on a stakeholder. | "What would you have done if they had said no?" |
| B4 | Tell me about a time you raised an ethics or fairness concern. | "How did you measure the problem?" |
| B5 | Tell me about a decision you made with incomplete data. | "Was it the right decision? How do you know?" |

### Set C — Project deep dive (technical storytelling)

| # | Question | Follow-up to expect |
|---|---|---|
| C1 | Give me the 2-minute pitch of your strongest project. | "Which part did you personally build?" |
| C2 | Why that metric, and what was your baseline? | "Is the difference bigger than noise?" |
| C3 | What didn't work? | "How quickly did you find out, and how?" |
| C4 | How would this change at 100× the data or traffic? | "What breaks first?" |
| C5 | What are you proudest of, and what would you change? | "Why didn't you do that the first time?" |

### Self-scoring rubric (0–2 per row, 10 points per answer)

| Dimension | 0 | 1 | 2 |
|---|---|---|---|
| **Structure** | Rambling; hard to follow | Mostly ordered, some backtracking | Clear STAR/CARL or six-beat pitch |
| **Specificity & ownership** | Generic; "we" throughout | One concrete moment, some "I" | Specific moment, clear personal actions and reasons |
| **Result** | No result | Vague result ("it went well") | Concrete result, with a number or a clear comparison when one exists |
| **Reflection** | None | Generic lesson | Specific lesson or change in behaviour since |
| **Time & delivery** | Over 2:30 or under 0:45 | 2:00–2:30, or rushed | 1:15–2:00, calm pace, stops to invite follow-ups |

**Reading your score:** 9–10 ready · 6–8 re-tell the story once more, fixing the lowest row · ≤ 5 rewrite the story in your bank first. Track your scores per story; after a week of drills your weakest *row* (often Result or Reflection) tells you what to practise.

## 📋 Answer Framework & Cheat Sheet

### STAR / CARL one-pager

| Part | Say | Share of a 2-min answer | Watch out for |
|---|---|---|---|
| **S**ituation / **C**ontext | Where, when, what was at stake (1–2 sentences) | ~15% | Over-explaining background |
| **T**ask | Your specific responsibility or goal | ~5% | Describing the team's goal only |
| **A**ction | 3–5 steps *you* took and **why** | ~50–60% | "We", skipping the reasoning |
| **R**esult | What changed; a number or comparison if real | ~15% | No result; inflated or unverifiable numbers |
| **L**earning | What you'd repeat or change | ~10% | Generic lessons ("communication is important") |

### 2-minute project pitch template

```
Problem      : {who} needed {outcome} because {pain}.
Constraints  : {data size / labels / latency / budget / privacy / deadline}.
Approach     : Baseline {simple thing} → {what I built} because {reason}.
Trade-offs   : I chose {X} over {Y}; the cost was {Z}.
Results      : {metric} = {my measured number} vs {baseline number} on {evaluation set}.
What I'd change: {honest limitation} → {next step}.
Close        : "Which part would you like me to go deeper on?"
```

### Résumé bullet formula

`Action verb` + `what you built, with the method` + `measured result vs a comparison` — ≤ 30 words, no tool lists, no numbers you can't defend.

### Drill-down question buckets

**Data** (source, size, labels) · **Splits & leakage** · **Metric** (why this one) · **Baseline** (compared to what) · **Failures** (what didn't work) · **Scale** (100× what breaks) · **Ownership** (what *you* did)

### ✅ Pre-interview checklist

- [ ] Story bank of 8–10 **true** stories, every competency covered by ≥ 2 (run `coverage_report`)
- [ ] Each story rehearsed out loud at least twice with a 2-minute timer
- [ ] "Tell me about yourself" (60–90 s) and "Why this company?" (specific reasons) prepared
- [ ] One project you can deep-dive for 45 minutes: baseline, metric, split, biggest failure, what you'd change
- [ ] Every number on your résumé: you know how it was measured and what it was compared against
- [ ] Every résumé bullet maps to a story (run `review_bullet` on each)
- [ ] Read the job description; mapped 3 requirements to 3 stories
- [ ] Researched the company's product and one recent engineering blog post or talk
- [ ] 3–5 questions to ask, tailored to each interviewer's role
- [ ] Logistics: time zone, video link, quiet room, water, charged laptop, notes with story titles (not scripts)
- [ ] Plan to send a short thank-you note within about a day

## 📚 Resources

### 📖 Behavioral interview guides
- [Tech Interview Handbook — Behavioral interviews for Software Engineers: How to prepare](https://www.techinterviewhandbook.org/behavioral-interview/) — STAR format, organising key stories, practising and mock interviews
- [Tech Interview Handbook — The 30 most common Software Engineer behavioral interview questions](https://www.techinterviewhandbook.org/behavioral-interview-questions/) — a practice list to run through your story bank
- [Tech Interview Handbook — Software Engineer interviews: Crafting the perfect self introduction](https://www.techinterviewhandbook.org/self-introduction/) — for "tell me about yourself"
- [Tech Interview Handbook — Software Engineer interviews: Best end of interview questions to ask](https://www.techinterviewhandbook.org/final-questions/) — questions grouped by technical work, role, culture, leadership, and company direction
- [MIT CAPD — Using the STAR method for your next behavioral interview (worksheet included)](https://capd.mit.edu/resources/the-star-method-for-behavioral-interviews/) — a university career office's STAR guide with a worksheet

### 📘 Books
- [Chip Huyen — Introduction to Machine Learning Interviews Book](https://huyenchip.com/ml-interviews-book/) — free online book on ML interviews from both sides of the table
  - [2.1.1.2 Non-technical skills](https://huyenchip.com/ml-interviews-book/contents/2.1.1.2-non-technical-skills.html) — what companies look for beyond technical ability
  - [2.3.1 Behavioral questions](https://huyenchip.com/ml-interviews-book/contents/2.3.1-behavioral-questions.html) — background, interests, communication, and personality questions in ML loops
  - [2.3.2 Questions to ask your interviewers](https://huyenchip.com/ml-interviews-book/contents/2.3.2-questions-to-ask-your-interviewers.html)

### 🏢 From hiring companies
- [Amazon — Leadership Principles](https://www.amazon.jobs/content/en/our-workplace/leadership-principles) — the principles Amazon's behavioral questions are built around; a good competency checklist even if you never apply there
- [Amazon — Interview Loop](https://amazon.jobs/content/en/how-we-hire/interview-loop) — how an interview loop is organised, with links on behavioral questions and the STAR method
- [Amazon — Applied Scientist Interview Prep](https://amazon.jobs/content/en/how-we-hire/applied-scientist-interview-prep) — official prep page for an ML-focused role

### 🧰 For your story bank and responsible-AI answers
- [Julia Evans — Get your work recognized: write a brag document](https://jvns.ca/blog/brag-documents/) — keep a running record of your work so stories and numbers are real and remembered
- [Mitchell et al. (2019) — Model Cards for Model Reporting](https://arxiv.org/abs/1810.03993) — a standard way to document a model's intended use, evaluation across groups, and limitations; great material for ethics and communication questions
- [Google PAIR — PAIR Guidebook](https://pair.withgoogle.com/guidebook/) — practical guidance on explaining AI behaviour and uncertainty to users

### 🏋️ Practice
- Use the three **Timed Drills** sets above with a friend, a study group, or a recording of yourself. Practising with a real person asking unscripted follow-ups is the closest thing to the real round.

## ➡️ What's Next

**[13 · End-to-End ML Project](../13_Capstone_Projects/01_End_to_End_ML_Project.ipynb)** — the best interview story is one you built and measured yourself. Build a capstone, go one step beyond the tutorial, then fill in the pitch template from section 4 with **your** numbers.

Then close the loop: return to **[01 · ML Coding Drills](01_ML_Coding_Drills.ipynb)** and cycle through the interview-prep notebooks — coding drills, theory, ML and LLM system design, and this storytelling round — until each timed drill feels routine.